packages/client/hmi-client/src/utils/pivot.ts

```diff
@@ -104,36 +103,53 @@ export const createMatrix1D = (data: any[]) => {
 		rows.push(row);
 	}
 
-	return { matrix: rows };
+	return rows;
 };
 
 export const createParameterMatrix = (
 	amr: Model,
 	transitionMatrixData: any[],
 	childParameterIds: string[]
 ) => {
-	let controllers: string[] = [];
 	let inputs: string[] = [];
 	let outputs: string[] = [];
 
 	const transitions = _.cloneDeep(
 		transitionMatrixData.map((t) => amr.model.transitions.filter(({ id }) => t.id === id)).flat()
 	);
 
+	const controllerIndexMap = new Map(); // Maps controllers to their input/output combo
+
 	// Get unique inputs and outputs and sort names alphabetically (these are the rows and columns respectively)
 	for (let i = 0; i < transitions.length; i++) {
 		const { input, output } = transitions[i];
 		// Extract and remove controllers out of inputs array
-		const newInputs = input.filter((ip: string) => !output.includes(ip));
-		const newOutputs = output.filter((ip: string) => !input.includes(ip));
+		const newInputs: string[] = [];
+		const newOutputs: string[] = [];
+		const newControllers: string[] = [];
+		for (let j = 0; j < input.length; j++) {
+			if (input[j] !== output[j]) {
+				newInputs.push(input[j]);
+				newOutputs.push(output[j]);
+			} else {
+				newControllers.push(input[j]);
+			}
+		}
+
+		if (!_.isEmpty(newControllers)) {
+			// Map controllers unique to these input/output combos
+			for (let j = 0; j < newInputs.length; j++) {
+				for (let k = 0; k < newOutputs.length; k++) {
+					controllerIndexMap.set(newInputs[j].concat('|', newOutputs[k]), newControllers);
+				}
+			}
```

Comment:
I don't thinks this is a cartesian product kind of situation. Probably more correct is soemthing like this

```
for (let j = 0; j < newInputs.length; j++) {
  const key = newInputs[j].concat('|', newOutputs[j])
  controllerIndexMap.set(...)
}
```